src/net_processing.cpp

```diff
@@ -4599,10 +4599,31 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
         // Start block sync
         if (pindexBestHeader == nullptr)
             pindexBestHeader = m_chainman.ActiveChain().Tip();
-        bool fFetch = state.fPreferredDownload || (nPreferredDownload == 0 && !pto->fClient && !pto->IsAddrFetchConn()); // Download if this is a nice peer, or we have no nice peers and this one might do.
+
+        // Determine whether we might try initial headers sync or parallel
+        // block download from this peer -- this mostly affects behavior while
+        // in IBD (once out of IBD, we sync from all peers).
+        bool sync_blocks_and_headers_from_peer = false;
+        if (state.fPreferredDownload) {
+            sync_blocks_and_headers_from_peer = true;
+        } else if (!pto->fClient && !pto->IsAddrFetchConn()) {
+            // Typically this is an inbound peer. If we don't have any outbound
+            // peers, or if we aren't downloading any blocks from such peers,
+            // then allow block downloads from this peer, too.
+            // We prefer downloading blocks from outbound peers to avoid
+            // putting undue load on (say) some home user who is just making
+            // outbound connections to the network, but if our only source of
+            // the latest blocks is from an inbound peer, we have to be sure to
+            // eventually download it (and not just wait indefinitely for an
+            // outbound peer to have it).
+            if (nPreferredDownload == 0 || mapBlocksInFlight.empty()) {
```

Comment:
nit: I would prefer combining this smth like 
```suggestion
        } else if (!pto->fClient && !pto->IsAddrFetchConn()) &&
                  (nPreferredDownload == 0 || mapBlocksInFlight.empty())) {
            // Typically this is an inbound peer. If we don't have any outbound
            // peers, or if we aren't downloading any blocks from such peers,
            // then allow block downloads from this peer, too.
            // We prefer downloading blocks from outbound peers to avoid
            // putting undue load on (say) some home user who is just making
            // outbound connections to the network, but if our only source of
            // the latest blocks is from an inbound peer, we have to be sure to
            // eventually download it (and not just wait indefinitely for an
            // outbound peer to have it).
```

Reply:
I prefer the current arrangement of the conditions, because:
1. The following comment says: `Typically this is an inbound peer.` This here is referring to the conditions `!pto->fClient && !pto->IsAddrFetchConn()` describing the node from which we want to fetch the block header. The other two conditions are specific to our node, which are checking if there is any preferred block to download on our side or if we have **Blocks in Flight**. Merging these two sets of conditions would make the following comment confusing.
2. Keeping these two sets of conditions separate makes it easier to understand the code’s logic and easier to follow with the comments.

Reply:
Thanks, I don't think many code reviewers will find the new arrangement of these conditions meaningfully harder to read than the old code here, and I do expect that many reviewers would indeed find the old code much harder to read, so I'm inclined to leave this as I've rewritten it.